File: app/day02/parts.py
import sys
from array import array

from app.support.utils import main, parse_module_to_day


def compute_diff(reports):
    return all(
        (
            (
                report - next_report <= 3
                if (next_report := reports[idx + 1 if idx < len(reports) - 1 else idx]) < report
                else next_report - report <= 3
            )
            for idx, report in enumerate(reports)
        ),
    )


def is_increasing(reports):
    return all(report > reports[idx + 1] for idx, report in enumerate(reports) if idx < len(reports) - 1)


def is_decreasing(reports):
    return all(report < reports[idx + 1] for idx, report in enumerate(reports) if idx < len(reports) - 1)


def is_safe(reports):
    return compute_diff(reports) and (is_increasing(reports) or is_decreasing(reports))


def compute_part_1(input_: str) -> int:
    return sum(is_safe(array("i", map(int, line.split()))) for line in input_.splitlines())


def compute_part_2(input_: str) -> int:
    amout_of_safe = 0
    for line in input_.splitlines():
        reports = array("i", map(int, line.split()))
        if is_safe(reports):
            amout_of_safe += 1
        else:
            for idx in range(0, len(reports)):
                if is_safe(reports[:idx] + reports[idx + 1 :]):
                    amout_of_safe += 1
                    break
    return amout_of_safe


if __name__ == "__main__":  # pragma: no cover
    module, day = parse_module_to_day(sys.modules[__name__].__package__)

    main(day, module, compute_part_1, compute_part_2)

File: app/day02/parts.js
const { readFile, strToInt } = require("../support/utils");

const computeDiff = (reports) =>
  reports.filter((report, idx) => {
    const nextReport = reports[idx < reports.length - 1 ? idx + 1 : idx];
    return nextReport < report
      ? report - nextReport <= 3
      : nextReport - report <= 3;
  }).length == reports.length;

const isIncreasing = (reports) =>
  !reports
    .map((report, idx) =>
      idx < reports.length - 1 ? report > reports[idx + 1] : true
    )
    .includes(false);

const isDecreasing = (reports) =>
  !reports
    .map((report, idx) =>
      idx < reports.length - 1 ? report < reports[idx + 1] : true
    )
    .includes(false);

const isSafe = (reports) =>
  computeDiff(reports) && (isIncreasing(reports) || isDecreasing(reports));

const compute_part_1 = () =>
  readFile("day02/input.txt")
    .split("\n")
    .map((line) => (line ? isSafe(line.split(/[\s]+/).map(strToInt)) : false))
    .reduce((acc, val) => (val ? acc + 1 : acc), 0);

const compute_part_2 = () =>
  readFile("day02/input.txt")
    .split("\n")
    .map((line) => {
      if (line) {
        const reports = line.split(/[\s]+/).map(strToInt);
        if (isSafe(reports)) {
          return true;
        } else {
          for (let index = 0; index <= reports.length; index++) {
            if (
              isSafe(reports.slice(0, index).concat(reports.slice(index + 1)))
            ) {
              return true;
            }
          }
        }
      }
      return false;
    })
    .reduce((acc, val) => (val ? acc + 1 : acc), 0);

console.assert(compute_part_1() == 341);
console.assert(compute_part_2() == 404);
